pkg/testmachinery/testflow/flow_operations.go

```diff
@@ -0,0 +1,162 @@
+package testflow
+
+import (
+	"fmt"
+	tmv1beta1 "github.com/gardener/test-infra/pkg/apis/testmachinery/v1beta1"
+	"github.com/gardener/test-infra/pkg/testmachinery/config"
+	"github.com/gardener/test-infra/pkg/testmachinery/locations"
+	"github.com/gardener/test-infra/pkg/testmachinery/testdefinition"
+	"github.com/gardener/test-infra/pkg/testmachinery/testflow/node"
+)
+
+// preprocessTestflow takes a Tesflow and creates a map which maps the unique step name to the step pointer.
+func preprocessTestflow(flowID FlowIdentifier, root *node.Node, tf tmv1beta1.TestFlow, loc locations.Locations, globalConfig []*config.Element) (map[string]*Step, map[*testdefinition.TestDefinition]interface{}, map[testdefinition.Location]interface{}, error) {
+	stepMap := make(map[string]*Step, 0)
+	testdefinitions := make(map[*testdefinition.TestDefinition]interface{}, 0)
+	usedLocations := make(map[testdefinition.Location]interface{}, 0)
+	for _, step := range tf {
+		// todo(schrodit): add validation
+
+		nodes, err := node.CreateNodesFromStep(step, loc, globalConfig, string(flowID))
+		if err != nil {
+			return nil, nil, nil, err
+		}
+
+		for n := range nodes {
+			testdefinitions[n.TestDefinition] = nil
+			usedLocations[n.TestDefinition.Location] = nil
+		}
+		stepMap[step.Name] = &Step{
+			Info:  step,
+			Nodes: nodes,
+		}
+	}
+	return stepMap, testdefinitions, usedLocations, nil
+}
+
+// CreateInitialDAG creates a DAG by evaluating the dependsOn steps.
+func CreateInitialDAG(steps map[string]*Step, root *node.Node) {
+	for _, step := range steps {
+		if step.Info.DependsOn == nil || len(step.Info.DependsOn) == 0 {
```

Comment:
```suggestion
		if step.Info.DependsOn == nil || len(step.Info.DependsOn) == 0 {
			addRootNodeAsParent(step)
		} else {
			addDependendsAsParent(step)
		}
```
Decouple a little more for better reading

Reply:
done for the `addDependentStepAsParent` function.
I think the first decoupling is unecessary as there are only 2 commands?

Reply:
This is also fine. It's just, if the code inside a function is consistently on the same abstraction level, it's better to read.